### PR TITLE
[WIP]Add ir emit support

### PR DESCRIPTION
**Pull request by splasky**
Add `--emit-lli` command for emit the AMaCC ir

### CODE REVIEW

File: amacc.c [hunk 1]
```diff
@@ -379,13 +420,13 @@ void next()
...
             while (*p != 0 && *p != tk) {
                 if ((ival = *p++) == '\\') {
-                    switch (ival = *p++) {
-                    case 'n': ival = '\n'; break; // new line
-                    case 't': ival = '\t'; break; // horizontal tab
-                    case 'v': ival = '\v'; break; // vertical tab
-                    case 'f': ival = '\f'; break; // form feed
-                    case 'r': ival = '\r'; break; // carriage return
-                    case '0': ival = '\0'; break; // an int with value 0
+                        switch (ival = *p++) {
```
Comment (jserv) [inline amacc.c]: Don't make style-oriented changes.

File: amacc.c [hunk 2]
```diff
@@ -52,6 +52,14 @@ int *n;              // current position in emitted abstract syntax tree
                      // emitted and pushed on the stack in the proper
                      // right-to-left order.
 int ld;              // local variable depth
+int ir,lli_fd;
+
+struct sym_table{
+    int type;
+    int val;
+    char* str;
```
Comment (jserv) [inline amacc.c]: Should be `char *str`.

File: vm/vm.c [hunk 3]
```diff
@@ -0,0 +1,348 @@
...
+        case DSYM:
+            r0 = dlsym(sp[1], *sp);
+            break;
+        case DLOP:
+            r0 = dlopen((char*)sp[1], *sp);
+            break;
+        case BSCH:
+            r0 = bsearch(sp[4], sp[3], sp[2], sp[1], *sp);
+            break;
+        case STRL:
```
Comment (jserv) [inline vm/vm.c]: How about eliminating `strlen` as an exposed function? That is, we can implement our own string length routine without declaring.

Comment (jserv) [reply vm/vm.c]: It can be `_strlen` or something alike.

**Comment by jserv** [comment]
Optional `-s` should generate the new IR instead, right?

**Comment by splasky** [comment]
`-s` option is for origin IR.
`--emit-ir` for generate new IR format.

**Comment by jserv** [comment]
> `-s` option is for origin IR.
> `--emit-ir` for generate new IR into output *.ll file.

Let's migrate to new IR.

**Comment by jserv** [comment]
It is silly to have two IR implementations.

**Comment by splasky** [comment]
`-s` options will generate IR with origin *.c information, `--emit-ir` will only generate ir for interpreter. Should we merge then together?

File: vm/vm.c (hunk 3 above)
Comment (splasky) [reply vm/vm.c]: > How about eliminating `strlen` as an exposed function? That is, we can implement our own string length routine without declaring.

Yes, we can. But is will generate more code length.  Change it?

**Comment by jserv** [comment]
> `-s` options will generate IR with origin *.c information, `--emit-ir` will only generate ir for interpreter. Should we merge then together?
> `-s` is also for debugging.

`--emit-ir` should generate IR assembly (so-called text form), which is unified with `-s` output (mixed with source code listing though).

File: vm/vm.c (hunk 3 above)
Comment (jserv) [reply vm/vm.c]: Implementing strlen only takes some lines of code.

File: runtest.py [hunk 4]
```diff
@@ -1,5 +1,7 @@
-#!/usr/bin/env python
-# encoding: utf-8
+#! /usr/bin/env python3
+# -*- coding: utf-8 -*-
+# vim:fenc=utf-8
+# Last modified: 2019-09-09 15:53:53
```
Comment (jserv) [inline runtest.py]: Don't insert date.

File: vm/util.h [hunk 5]
```diff
@@ -0,0 +1,106 @@
+#ifndef UTIL_H_G3U7SPJM
```
Comment (jserv) [inline vm/util.h]: Revise the identifier.

File: amacc.c [hunk 6]
```diff
@@ -226,17 +235,54 @@ enum {
...
+        memset(buf,'\0',len);
+        int k = snprintf(buf, len, "%%%d = %4.4s", E,
+           & "LEA  IMM  JMP  JSR  BZ   BNZ  ENT  ADJ  LEV  "
+            "LI   LC   SI   SC   PSH  "
+            "OR   XOR  AND  EQ   NE   LT   GT   LE   GE   "
+            "SHL  SHR  ADD  SUB  MUL  "
+            "OPEN READ WRIT CLOS PRTF MALC FREE "
+            "MSET MCMP MCPY MMAP "
+            "DSYM BSCH STRT DLOP DIV  MOD  SNPR EXIT CLCA" [*E * 5]);
+        if(src){
```
Comment (jserv) [inline amacc.c]: Use consistent style.

**Comment by jserv** [comment]
The pull request conflicts with the latest `master` branch. Let's drop this and see if we can rework later.

**Pull request closed without merging**